lib/vnet/app_resolver.go

```diff
@@ -0,0 +1,178 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package vnet
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	apiclient "github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// AppProvider is an interface providing the necessary methods to log in to apps and get clients able to list
+// apps in all clusters in all current profiles. This should be the minimum necessary interface that needs to
+// be implemented differently for Connect and `tsh vnet`.
+type AppProvider interface {
```

Comment:
yeah, i mean it's hard to name since it does a few things, even later on it provides the cluster `vnet_config` as well. I guess I named it this way since the most important thing it does is provide the necessary means to list apps in multiple clusters and get certs for them. for now i'll probably leave this if only to reduce conflicts